torch/csrc/jit/python/python_sugared_value.h

```diff
@@ -286,5 +300,121 @@ struct VISIBILITY_HIDDEN PythonClassValue : public ClassValue {
   py::object py_type_;
 };
 
+struct VISIBILITY_HIDDEN ModuleDictMethodRecursive : public SugaredValue {
+  explicit ModuleDictMethodRecursive(
+      std::shared_ptr<ModuleValue> module,
+      std::shared_ptr<SugaredTupleValue> keys,
+      std::shared_ptr<SugaredTupleValue> values,
+      const std::string& name)
+      : module_(module), keys_(keys), values_(values), name_(name){};
+
+  std::string kind() const override {
+    return name_;
+  }
+
+  std::shared_ptr<SugaredValue> call(
+      const SourceRange& loc,
+      Function& f,
+      at::ArrayRef<NamedValue> inputs,
+      at::ArrayRef<NamedValue> attributes,
+      size_t n_binders) override {
+    auto iterator = std::make_shared<IterableTree>();
+
+    std::map<std::string, NamedValue> inputMap;
+    for (const auto& input : inputs) {
+      if (input.hasName()) {
+        inputMap.insert({input.name(), input});
+      } else {
+        throw ErrorReport(loc)
+            << "Please explicitly name all parameters passed to: " << name_
+            << "()";
+      }
+    }
+
+    if (inputMap.find("prefix") != inputMap.end()) {
+      throw ErrorReport(loc)
+          << "Prefix is not currently supported yet for: " << name_ << "()";
+    }
+
+    if (inputMap.find("recurse") != inputMap.end()) {
+      auto recurseInputValue =
+          inputMap.find("recurse")->second.value(*f.graph());
+      if (recurseInputValue->type()->kind() == TypeKind::BoolType) {
+        bool shouldRecurse = constant_as<bool>(recurseInputValue).value();
+        if (shouldRecurse) {
+          std::vector<SugaredValuePtr> moduleKeys;
+          std::vector<SugaredValuePtr> moduleValues;
+
+          std::vector<SugaredValuePtr> iterValues;
+          std::vector<SugaredValuePtr> iterKeys;
+
+          auto lambda = [&](std::shared_ptr<ModuleValue> m) -> void {
```

Comment:
`moduleKeys` and `iterKeys` seem sort of redundant here. They both have a very similar purpose: containers that are populated as we descend through the module hierarchy. I think it would be a good cleanup to refactor `recurseThroughNestedModules()` to remove the `keys` and `values` arguments. That way it implements just the traversal, and defers to the visitor lambda to actually set up whatever structures the caller expects.

Reply:
Yeah super interesting. Let me play with what that would look like.